Add preprocessor for idlc

k0ekk0ek opened the pull request:
This PR adds an mcpp 2.7.2 (as per CQ# 18134) based preprocessor for the IDL compiler that is currently being developed. It consists of the mcpp 2.7.2 sources submitted to IPZilla, newly written build files and a commit that fixes the compiler warnings in the mcpp sources.

mcpp is forked mainly because the author of the upstream project is inactive, though the added `configed.H.in` header tailors mcpp more towards a preprocessor for `idlc`, whereas normally mcpp tries to imitate the resident preprocessor as much as possible.

I wouldn't spend too much time reviewing the commit that adds the sources for mcpp 2.7.2, but it would be good to evaluate if you agree with how the preprocessor is configured at compile time, e.g. the macro `__CYCLONEDSS__` is exported by default and if I didn't make any errors in the warning fixes.

eboasson commented:
I had a somewhat longer look at the changes for fixing the warnings. Most seen straightforward, and most of the others appear reasonable — but there's always a risk in changing signed types to unsigned types, and even more so changing ``int``s to ``char``s (especially considering that the signedness of ``char`` is not a given). I've done a few sanity checks, just to be sure.

It turns out Xcode (so presumably Clang too with all the warnings enabled) still has another few to nag about, as does MSVC. In my view, the way to deal with this is either to address all the warnings and be sure the changes are correct, or to carefully inspect the original code and determine that the warnings are nuisance ones.

The former is obviuosly attractive: even though verifying correctness does not appear to be a trivial exercise, it in the end it is always worth it. The flip side is that there is reason to believe the existing code to be correct (and certainly most of the warnings I still see in Xcode are evidently nuisance ones, like the "possible misuse of comma operator") and that the time might be better invested elsewhere for now. That line of reasoning would lead one to simply suppress some warnings for these files.

In my view, they are both reasonable positions under the circumstances. Feel free to pick your poison :)

k0ekk0ek commented:
@eboasson, I'm going to try and fix the warnings. I've been able to use `snprintf` in a couple of places instead of plain `sprintf`. For Windows I added `_CRT_SECURE_NO_WARNINGS`, which seems reasonable, at least for the short-term. Apart from that I synchronized the warning flags between Xcode and plain Clang, so we at least see the same warnings.

When the warnings are gone, I'm going to run the modified sources over the test-cases that ship with the original mcpp. That way, we know it's at least passing the tests of the original author and that should get us on-par (more-or-less) with the original v2.7.2 of mcpp. I guess I'll go over the changes one more time after that before issuing a "final" PR.

Of course, the other comments will also be addressed. Thanks!

k0ekk0ek commented:
* The target system must be known for some platform specific things. Windows for backslashes and the environment separator among other things. I could leave out the platform macros, but I figured they might also come in handy. It's never going to be compiled for embedded targets, so the list won't really grow that much. I've added the `__linux` macro.
* DDS_VERSION indeed seems a little to vague. I was thinking just `CYCLONEDDS`, as its for IDL, not C(?)
* I figured the checks are not in the way, so I'd like to leave them in for now. They might prove to be good examples.

Well the warnings are gone. I enabled some more warnings for Clang while I was at it and added `DDSRT_WARNING_CLANG_OFF` and `DDSRT_WARNING_CLANG_ON` macros to disable `assign-enum` in `ddsc/tests/subscriber.c`.

Still need to run the test suite and go over the changes one more time though.

k0ekk0ek commented:
@eboasson, I fixed a couple of warnings that I got after running `sparse` and adding the `printf` function attribute to `mcpp_fprintf` and went over the warning fixes one more time.

I also ran the tests that come with `mcpp` that I could. Turns out `stdarg.h` isn't actually a header. At least on my system, so I made some modifications. Most of them ran fine, some others I couldn't run because those modification lead to redefinitions. To be fair though the output of `mcpp` looked just fine, it's just that I couldn't compile the program because I needed to include `stdio.h` after preprocessing with `mcpp`.

I changed `__CYCLONEDDS__` to `CYCLONEDDS` for now, figuring we can rename it later on. Also, I removed the default include directory for now as we don't actually ship default `.idl` files.

All-in-all I think this PR in a very decent state and ready to be merged(?)

P.S.
I'm very sorry for the force pushes, I could have done that a bit better, but for Windows there was a lot of trial and error, fixing things in the first commits then fixing some warnings... I'll do my best to streamline that better next time. If you want to go over what changed and you still have my old branch checked out, may I suggest running a plain `diff` between the two.

eboasson commented:
Hi @k0ekk0ek, no worries about having had to rewrite the history. It looks ok to me, but there are few details I’d like you to consider before merging it.

* ``CheckTargetCPU.c.in``: could do with the addition of ARM.
* There are some warnings when I compile it on macOS for a release build, so one would think there is still some difference in set of enabled warnings. See the list below, it looks to me like these are all old, so I don’t see why they would have to be addressed in this PR (I apologise for not checking as thoroughly last time):
  * Clang has a lovely feature for checking doxygen comments and it complains about a bunch of them;
  * there two missing prototypes;
  * and a bunch of (what I think are) nuisance warnings;
  * Clang’s static analyzer gives more still, probably nuisance warnings, but still worth looking into (it looks way worse than it is). Some originate in mcpp, but there are also old ones in ``ddsts`` (note that it makes little sense to run the analyzer with assertions compiled out because those assertions are often the best way to deal with false positives).

My suggestion would be to at least add ARM and address the remaining warnings in the new bits. Given that the status of the ``xtypes`` branch, that would be surely be sufficient at this point. And then, please just go ahead and merge it.

```
=== ANALYZE TARGET ddsts OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
In file included from /Users/erik/adlink/cyclonedds-2/src/ddsts/src/type_walk.c:16:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
7 warnings generated.
=== ANALYZE TARGET ddsts OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
In file included from /Users/erik/adlink/cyclonedds-2/src/ddsts/src/typetree.c:17:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
7 warnings generated.
=== ANALYZE TARGET idlpp OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/eval.c:259:29: warning: this function declaration is not a prototype
      [-Wstrict-prototypes]
    { T_FPTR,   0,          S_PFPTR     },          /* int (*())    */
                            ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/eval.c:197:37: note: expanded from macro 'S_PFPTR'
#define S_PFPTR     (sizeof (int (*)()))
                                    ^
1 warning generated.
=== ANALYZE TARGET idlpp OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:592:42: warning: variable 'cur_edge' may be uninitialized when used here
      [-Wconditional-uninitialized]
                                memmove( cur_edge, cur_edge + len
                                         ^~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:40:
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.H:45:
In file included from /Users/erik/adlink/cyclonedds-2/XX/src/tools/idlpp/src/configed.H:297:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/string.h:190:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/secure/_string.h:69:28: note: 
      expanded from macro 'memmove'
                __builtin___memmove_chk (dest, __VA_ARGS__, __darwin_obsz0 (dest))
                                         ^~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:551:41: note: initialize the variable 'cur_edge' to silence this warning
                    char *      cur_edge;
                                        ^
                                         = NULL
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1507:36: warning: variable 'mark' may be uninitialized when used here
      [-Wconditional-uninitialized]
            rm = mac_id[ mac_n++][ mark];
                                   ^~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1474:30: note: initialize the variable 'mark' to silence this warning
        size_t  num = 0, mark;
                             ^
                              = 0
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:2491:21: warning: variable 'loc' may be uninitialized when used here
      [-Wconditional-uninitialized]
                    loc++;
                    ^~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:2446:20: note: initialize the variable 'loc' to silence this warning
    LOCATION *  loc;                            /* Current locs     */
                   ^
                    = NULL
3 warnings generated.
=== ANALYZE TARGET idlpp OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:1661:21: warning: declaration shadows a variable in the global scope
      [-Wshadow]
    const char *    mcpp_optarg,
                    ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:233:17: note: previous declaration is here
static char *   mcpp_optarg;
                ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:2379:21: warning: declaration shadows a variable in the global scope
      [-Wshadow]
    int             framework       /* Setting a framework directory*/
                    ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:352:17: note: previous declaration is here
static char *   framework[ MAX_FRAMEWORK];  /* Framework directories*/
                ^
2 warnings generated.
=== ANALYZE TARGET idlpp OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/directive.c:1418:23: warning: Null pointer argument in call to string length
      function
        s_parmnames = strlen( parmnames) + 1;
                      ^~~~~~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/directive.c:1421:14: warning: Null pointer argument in call to string length
      function
    s_repl = strlen( repl) + 1;
             ^~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/directive.c:1608:41: warning: Null pointer argument in call to string length
      function
                        parms[ i].len = strlen( cp);
                                        ^~~~~~~~~~~
3 warnings generated.
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/eval.c:434:21: warning: Undefined or garbage value returned to caller
                    return  valp[-1].val;   /* Finished ok.         */
                    ^~~~~~~~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/eval.c:1301:8: warning: Assigned value is garbage or undefined
    v1 = (--valp)->val;
       ^ ~~~~~~~~~~~~~
2 warnings generated.
=== ANALYZE TARGET idlpp OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1019:36: warning: Array access (from variable 'arglist') results in a
      null pointer dereference
            out = stringize( defp, arglist[ c], out);
                                   ^~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1024:17: warning: Dereference of null pointer
      (loaded from variable 'prev_token')
            if (*prev_token == DEF_MAGIC || *prev_token == IN_SRC) {
                ^~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1053:20: warning: Dereference of null pointer
            *out++ = MAC_PARM;
            ~~~~~~~^~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1075:10: warning: Dereference of null pointer
      (loaded from variable 'out')
    *out = EOS;         /* Ensure terminatation in case of no token */
     ~~~ ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1105:16: warning: Array access (from variable 'arglist') results in a
      null pointer dereference
        argp = arglist[ c];                 /* Actual argument      */
               ^~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1154:32: warning: Array access (from variable 'arglist') results in a
      null pointer dereference
        out = stringize( defp, arglist[ c], out);
                               ^~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1158:16: warning: Array access (from variable 'arglist') results in a
      null pointer dereference
        argp = arglist[ c];                 /* Actual argument      */
               ^~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1517:26: warning: Array access (from variable 'loc_tab') results in a
      null pointer dereference
            memmove( tp, loc_tab[ num], len);
                         ^~~~~~~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:40:
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.H:45:
In file included from /Users/erik/adlink/cyclonedds-2/XX/src/tools/idlpp/src/configed.H:297:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/string.h:190:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/secure/_string.h:69:34: note: 
      expanded from macro 'memmove'
                __builtin___memmove_chk (dest, __VA_ARGS__, __darwin_obsz0 (dest))
                                               ^~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:1972:46: warning: Branch condition evaluates to a garbage value
                    if (inner->nargs >= 0 && mgc_seq.magic_start) {
                                             ^~~~~~~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:2491:21: warning: The expression is an uninitialized value. The computed
      value will also be garbage
                    loc++;
                    ^~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:2496:17: warning: The expression is an uninitialized value. The computed
      value will also be garbage
                loc++;                  /* Advance pointer to infs  */
                ^~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:2587:40: warning: The left operand of '-' is a garbage value
                        , (size_t)(loc - locs) * sizeof (LOCATION));
                                   ~~~ ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:2646:37: warning: Dereference of undefined pointer value
                (*locp)->start_line = s_line_col.line;
                ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/expand.c:2803:13: warning: The expression is an uninitialized value. The computed
      value will also be garbage
            (*locp)++;      /* Advance pointer even if !trace_arg   */
            ^~~~~~~~~
14 warnings generated.
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/support.c:1745:48: warning: The left operand of '&' is a garbage value
            } else if (! (char_type[ *(tp - 1) & UCHARMAX] & HSP)) {
                                     ~~~~~~~~~ ^
1 warning generated.
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:526:17: warning: Null pointer argument in call to string comparison
      function
            if (str_eq( mcpp_optarg, "post")
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:43:
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/internal.H:305:26: note: expanded from macro 'str_eq'
#define str_eq(s1, s2)  (strcmp(s1, s2) == 0)
                         ^~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:739:17: warning: Dereference of null pointer
            if (*(mcpp_optarg + 1) == EOS && isdigit( *mcpp_optarg))
                ^~~~~~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:753:17: warning: Null pointer argument in call to string comparison
      function
            if (str_eq( mcpp_optarg, "-")) {        /* -I-                  */
                ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:43:
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/internal.H:305:26: note: expanded from macro 'str_eq'
#define str_eq(s1, s2)  (strcmp(s1, s2) == 0)
                         ^~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:879:17: warning: Null pointer argument in call to string comparison
      function
            if (str_eq( mcpp_optarg, "M")) {                /* -MM  */
                ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:43:
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/internal.H:305:26: note: expanded from macro 'str_eq'
#define str_eq(s1, s2)  (strcmp(s1, s2) == 0)
                         ^~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:910:17: warning: Null pointer argument in call to string comparison
      function
            if (str_eq( mcpp_optarg, "64")) {               /* -m64 */
                ^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:43:
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/internal.H:305:26: note: expanded from macro 'str_eq'
#define str_eq(s1, s2)  (strcmp(s1, s2) == 0)
                         ^~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:1058:17: warning: Dereference of null pointer
      (loaded from variable 'mcpp_optarg')
            i = *mcpp_optarg;
                ^~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:1209:26: warning: Dereference of null pointer
      (loaded from variable 'mcpp_optarg')
            if (isdigit( *mcpp_optarg)) {
                         ^~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:2486:5: warning: Null pointer argument in call to string copy function
    strcpy( slbuf1, dir);                   /* Include directory    */
    ^~~~~~~~~~~~~~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:42:
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.H:45:
In file included from /Users/erik/adlink/cyclonedds-2/XX/src/tools/idlpp/src/configed.H:297:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/string.h:190:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/secure/_string.h:83:3: note: 
      expanded from macro 'strcpy'
                __builtin___strcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:3616:16: warning: Null pointer argument in call to string length function
    fnamelen = strlen( filename);
               ^~~~~~~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:4670:16: warning: Dereference of null pointer
      (loaded from variable 'colon')
        *colon = EOS;
         ~~~~~ ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:4683:9: warning: Null pointer argument in call to string copy function
        strcpy( comment - 2, "\n");         /* Remove the comment   */
        ^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.c:42:
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlpp/src/system.H:45:
In file included from /Users/erik/adlink/cyclonedds-2/XX/src/tools/idlpp/src/configed.H:297:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/string.h:190:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/secure/_string.h:83:3: note: 
      expanded from macro 'strcpy'
                __builtin___strcpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11 warnings generated.
=== ANALYZE TARGET idlc-lib OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
In file included from src/idl.l:19:
In file included from src/idl.y:32:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1789:11: warning: parameter 'bytes' not found in the function
      declaration [-Wdocumentation]
 * @param bytes the byte buffer to scan
          ^~~~~
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1790:11: warning: parameter 'len' not found in the function
      declaration [-Wdocumentation]
 * @param len the number of bytes in the buffer pointed to by @a bytes.
          ^~~
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1877:5: warning: no previous prototype for function
      'ddsts_parser_get_column' [-Wmissing-prototypes]
int ddsts_parser_get_column  (yyscan_t yyscanner)
    ^
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1935:21: warning: empty paragraph passed to '@param' command
      [-Wdocumentation]
 * @param line_number
   ~~~~~~~~~~~~~~~~~^
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1950:21: warning: empty paragraph passed to '@param' command
      [-Wdocumentation]
 * @param line_number
   ~~~~~~~~~~~~~~~~~^
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1950:11: warning: parameter 'line_number' not found in the function
      declaration [-Wdocumentation]
 * @param line_number
          ^~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1950:11: note: did you mean 'column_no'?
 * @param line_number
          ^~~~~~~~~~~
          column_no
/Users/erik/adlink/cyclonedds-2/XX/src/tools/idlc/src/idl.lexer.c:1953:6: warning: no previous prototype for function
      'ddsts_parser_set_column' [-Wmissing-prototypes]
void ddsts_parser_set_column (int  column_no , yyscan_t yyscanner)
     ^
14 warnings generated.
In file included from src/idl.y:20:
In file included from src/idl.y:32:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
7 warnings generated.
=== ANALYZE TARGET idlc-lib OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/gen_c99.c:25:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/gen_c99.c:1744:25: warning: declaration shadows a local variable [-Wshadow]
          ddsts_type_t *struct_def = get_struct_def(member->declaration.decl_type);
                        ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/gen_c99.c:1726:17: note: previous declaration is here
  ddsts_type_t *struct_def = get_struct_def(path->type);
                ^
8 warnings generated.
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/parser.c:20:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
7 warnings generated.
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/tt_create.c:21:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
/Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/tt_create.c:783:26: warning: declaration shadows a local variable [-Wshadow]
      for (ddsts_type_t *member = struct_def->struct_def.members.first; member != NULL && declaration == NULL; member = member...
                         ^
/Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/tt_create.c:752:19: note: previous declaration is here
    ddsts_type_t *member = NULL;
                  ^
8 warnings generated.
=== ANALYZE TARGET idlc-lib OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
/Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/gen_c99.c:946:3: warning: Value stored to 'rc' is never read
  rc = ddsts_declaration_is_key(declaration, &declaration_is_key);
  ^    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
=== ANALYZE TARGET idlc OF PROJECT CycloneDDS WITH CONFIGURATION Debug ===
In file included from /Users/erik/adlink/cyclonedds-2/src/tools/idlc/src/main.c:18:
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:193:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_array_t struct
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:298:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] module  A non-NULL pointer to a ddsts_module_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:368:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:381:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] struct_def  A non-NULL pointer to a ddsts_struct_t struct.
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:10: warning: unrecognized parameter passing direction,
      valid directions are '[in]', '[out]' and '[in,out]' [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
         ^~~~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:420:18: warning: parameter 'array' not found in the function
      declaration [-Wdocumentation]
 * @param[inout] array         A non-NULL pointer to an ddsts_declaration_t
                 ^~~~~
/Users/erik/adlink/cyclonedds-2/src/ddsts/include/dds/ddsts/typetree.h:422:18: warning: parameter 'element_type' not found in the
      function declaration [-Wdocumentation]
 * @param[in]    element_type  A non-NULL pointer to the declaration type. If the
                 ^~~~~~~~~~~~
7 warnings generated.
```

k0ekk0ek commented:
Thanks @eboasson!

I've addressed the remaining points.

* Support for ARM added.
* Warnings are fixed. Luckily I could bother someone for a macOS machine, which helped a lot for the reports from Xcode's static analyzer.
  * I've added `-Wdocumentation` and addressed the issues in `ddsts` and disabled them for the `flex` generated sources.
  * The missing prototypes were also in the `flex` generated sources, so I've added `-Wno-missing-prototypes` there.
  * All gone, at least on all the machines that I could test on.
  * I went through the reports by the static analyzer and tried to carefully reproduce, but I found all of them to be false-positives. `assert` was added in a couple of cases, in a couple of others initializing the variable was a good solution. I've modified some actual code in one occasion too.
  * Renamed a generated variable in the CUnit header files to fix some shadow warnings.

I repeated my tests with the test-cases that come with the original version of `mcpp` and they ran just fine. The Debian project has some additional patches for the preprocessor, I've issued a CQ to include those, but I'll create a new PR for that.

If the builds pass, I'm pressing merge tomorrow...
